model/common/src/icon4py/model/common/orchestration/decorator.py

```diff
@@ -0,0 +1,673 @@
+# ICON4Py - ICON inspired code in Python and GT4Py
+#
+# Copyright (c) 2022-2024, ETH Zurich and MeteoSwiss
+# All rights reserved.
+#
+# Please, refer to the LICENSE file in the root directory.
+# SPDX-License-Identifier: BSD-3-Clause
+
+"""
+DaCe Orchestration Decorator
+
+Creates a DaCe SDFG that fuses any GT4Py Program called by the decorated function.
+"""
+
+from __future__ import annotations
+
+import hashlib
+import inspect
+import os
+import shutil
+import uuid
+from collections.abc import Callable
+from pathlib import Path
+from types import ModuleType
+from typing import Any, Optional, Union, get_type_hints
+
+import gt4py.next as gtx
+import numpy as np
+from gt4py._core import definitions as core_defs
+
+from icon4py.model.common import dimension as dims, settings
+from icon4py.model.common.decomposition import definitions as decomposition
+from icon4py.model.common.grid import icon as icon_grid
+from icon4py.model.common.orchestration import dtypes as orchestration_dtypes
+
+
+try:
+    import dace
+except ImportError:
+    dace: Optional[ModuleType] = None  # type: ignore[no-redef]
+
+try:
+    import ghex
+    from ghex import expose_cpp_ptr
+except ImportError:
+    ghex: Optional[ModuleType] = None  # type: ignore[no-redef]
+
+try:
+    import mpi4py
+    from mpi4py import MPI
+except ImportError:
+    mpi4py: Optional[ModuleType] = None
+
+if dace:
+    from dace import hooks
+    from dace.transformation.passes.simplify import SimplifyPass
+    from gt4py.next.program_processors.runners.dace_common.utility import (
+        connectivity_identifier,
+    )
+
+
+def orchestrate(func: Callable | None = None, *, method: bool | None = None):
+    def _decorator(fuse_func: Callable):
+        if settings.dace_orchestration is not None:
+            local_cache = {}  # Caching
+
+            if "dace" not in settings.backend.name.lower():
+                raise ValueError(
+                    "DaCe Orchestration works only with DaCe backends. Change the backend to a DaCe supported one."
+                )
+
+            self_name = next(iter(inspect.signature(fuse_func).parameters))
+
+            # If not explicitly set by the user, assume the provided callable is a method
+            # when its first argument is called 'self'
+            func_is_method = method or (self_name == "self")
```

Comment:
There is a simpler and more robust way of checking for "method-ness":
```suggestion
            # If not explicitly set by the user, check the argument is a method
            func_is_method = method or isinstance(fuse_func, types.MethodType)
```

Note: this of course requires `types` (a builtin) to be imported.